python/pyarrow/_dataset.pyx

```diff
@@ -2746,6 +2746,10 @@ cdef class Scanner(_Weakrefable):
     use_threads : bool, default True
         If enabled, then maximum parallelism will be used determined by
         the number of available CPU cores.
+    use_async : bool, default False
+        If enabled, the an async scanner will be used that should offer
+        better performance with high-latency/highly-parallel filesystems
+        (e.g. S3)
```

Comment:
The option needs to be added to _populate_builder and Scanner.from_fragment/Scanner.from_dataset or else it won't actually take effect.

Reply:
Should we also add the parameter to the tests? Maybe some refactoring is needed though to make it easier to share parameters like this.

Reply:
I updated all the tests so that they are parameterized on use_threads/use_async.  This helped identify the spots in _populate_builder, etc. where I had missed passing the flag through.

Reply:
I do think we might want to reduce the number of places we pass parameters around but this was a problem before use_async (e.g. there are a lot of places in the docs where we say "See scan method parameters documentation.").  That's not for this PR though and style-wise it might not be so bad so python can just forward kwargs but when it dips into cython it seems we need to materialized all the kwargs.

Reply:
It's because `_populate_builder` is cdef, i.e. gets compiled into an actual C function with a C parameter list, so it can't take `**kwargs` (but can directly take things like a shared_ptr). Though, we could pack all the non-C(++) arguments into a single `kwargs`.

Reply:
> Though, we could pack all the non-C(++) arguments into a single kwargs.

I don't know what you mean by this.  Do you mean taking all these arguments and putting them into some kind of `ScanOptions` class?

Reply:
Ah, sorry, what I mean is that if you're concerned about expanding all the parameters everywhere, you could declare 

```
cdef _populate_builder(shared_ptr[CScannerBuilder]& builder, object kwargs):
    # Extract args from kwargs individually...
```

and then make every user-facing method take just `**kwargs`. Of course, that just pushes the boilerplate around a bit (e.g.potentially having to cast things inside `_populate_builder`).

Doesn't need to be done here, just wanted to mention one way to reduce the number of places where we list all the parameters.